### ISSUE TITLE
How to jsdoc functions as parameters

### ISSUE THREAD

**Issue by zepumph** [phetsims/phet-info#107]
I think it is likely that we have discussed this before, but I can't seem to find a consensus. From this issue I would like to get an idea of how PhET devs want to annotate functions in jsdoc. I would then like to make sure that ends up in the coding conventions section of the code review checklist.

To me this applies to functions passed into other functions, functions passed in as options, and functions returned from functions. Each time it feels awkward and incomplete.

Here are some ways I've seen it done before:

Example 1 - the simple way. No need to communicate any more info. Often this is enough for me
```js
/**
 * @param {function} callback
 */
x = function( callback) { callback() };
```

Example 2 - plain english. If needing to describe more about the function, we just use words. This can get ugly quickly, and is hard to parse because all devs write it a bit differently.

```js
/**
 * @param {function} callback - called with {string} of your favorite ice cream as the first parameter
 */
x = function( callback) { callback( 'chocolate, duh' ) };
```

Example 3 - google closure syntax that I've seen before, but haven't really used it but it seems really nice. See here https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System. It is a nice way to communicate parameter/return types for anonymous functions, but it is hard to describe them.

```js
/**
 * @param {function(string)} callback  
 */
x = function( callback) { callback( 'still chocolate' ) };

// and for return types too:
/**
 * @param {function(string):number} stringToNum 
 */
x = function( stringToNum) { const myNum = stringToNum( 'chocolate. . . mmmm' ) };
```

Example 4: use `@name`, `@callback`, or `@typedef` annotations in a separate block to define the inner parameter. This is a bit clunky, and I think that we have nixed this before.

```js
/**
 * @name mySpecialCallback
 * Converts a string to a number
 * @param {string}
 * @returns {number}
 */
/**
 * @param {mySpecialCallback} callback  
*/
x = function( callback) { callback( 'still chocolate' ) };
```

My personal priorities here are:
* Have a clear way of doing things
* Have that way documented

I really like the idea of example 3. IDEA uses google closure syntax, so we could get more out of our ide (untested), and it is clear and concise. Marking for dev meetings to see others' opinions.

**Comment by pixelzoom** [comment]
I generally treat anonymous functions like named functions or methods.  But if I'm providing a callback to a some other method, then the callback signature should be part of the method's documentation, not the callbacks.  

For example, I feel no need to document the callback instances in this case:

```js
// Screen.js
  /**
   * @param {function} createModel
   * @param {function:Object } createView - function( model )
   * @param {Object} [options]
   * @constructor
   */
  function Screen( createModel, createView, options ) {

// some Screen created in a main.js
... = new Screen( 
  () => new MyMode(),
  model => new MyScreenView( model ),
  ...
);
```

**Comment by samreid** [comment]
* I don't recall nixing Example 4 before, and was interested to see that IDEA knows how to navigate it properly and it has the advantage that it matches our existing JSDoc strategy.  However, it seems odd (but idiomatic) to say `@param {mySpecialCallback}` instead of `@param {function} of type mySpecialCallback` though.
* In `@param {function} createModel` and `@param {function:Object } createView - function( model )` it seems `function:Object` has `Object` referring to the argument type, but not the return type?   Is that intentional?  Why are the function parameters not documented?
* One day it may be nice to use machine-readable JSDoc for automatic type checking, navigation and search like https://github.com/Kegsay/flow-jsdoc

**Comment by zepumph** [comment]
We discussed this today during dev meeting, and came up with the following proposal and plan to move forward. We came up with a sort of sliding scale based on the complexity that is needed. Note that as part of this we no longer accept `{function}` as acceptable documentation. It is too vague.

1. Most basic option it to use Google Closure Type syntax, for more info see https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System. This specifies the param/return types, but nothing more here are some examples:
    * `@param {function(number, string):Node} getNode`
    * `@param {function(number, number):Vector2} getVector2`
    * `@param {function():(number|boolean)} randomlyReturnNumberOrBoolean`
    * `@param {function():(string|null)} getStringAndSometimesNull`
    * `@param {function(number)} giveMeNumberAndReturnNothing`
    * `@param {function()} noParamsAndNoReturnValue`

2. When needing to be a bit more specific, we can use this syntax to name the parameters of the function. Sometimes this is all that is needed for clarity on what the param does: 
    * `@param {function(model:MyModel, length:number, name:string): Node} getLengthNode`
    * `@param {function(aSelfExplanatoryNameForAString:string): Node} getStringNode`
3. Finally, if this isn't enough, use english to explain the parameters and return values. This is easy because they are named, and can be easily mentioned:
    * `@param {function(model:MyModel, length:number, name:string): Node} getLengthNode - returns the length Node that you have always wanted, name is the name, length is a special number according to the following 24 criteria. . .`

4. If using jsdoc rendering tools (like phet-io does), you must use a compatible format (not 2 or 3), and you may need to use the more complicated solution:
	```js
	/**
	 * @name mySpecialCallback
	 * Converts a string to a number
	 * @param {string}
	 * @returns {number}
	 */
	/**
	 * @param {mySpecialCallback} callback  
	*/
	x = function( callback) { callback( 'still chocolate' ) };
	```

Note that when testing JSDoc rendering, options 1-3 were all compiled down into just `{function}`, which is actually alright with me, at least for now. At least it didn't break.


I updated PERENNIAL/Maintenance.js to use this new format. From here I think the steps should be:
  * Have everyone review the proposal, both above and in Maintenance.js.
  * Discuss at a dev meeting where we have a solid quorum.
  * Document our decision in code conventions section of the code review checklist.

**Comment by chrisklus** [comment]
From 08/22/19 dev meeting - no objections! This sliding scale sounds great. @zepumph is going to add a condensed version of this to the code review checklist.
